Crash report (develop): NPE in UploadStore.handleUpdateMedia

Steps:

1. Create a new post
2. Go to post settings
3. Tap "Set featured image" button
4. Pick an image from the local library -> 💥 

```
  Process: org.wordpress.android.beta, PID: 6959
  org.greenrobot.eventbus.EventBusException: Invoking subscriber failed
      at org.greenrobot.eventbus.EventBus.handleSubscriberException(EventBus.java:505)
      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:487)
      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479)
      at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
      at java.lang.Thread.run(Thread.java:761)
   Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int org.wordpress.android.fluxc.model.MediaModel.getId()' on a null object reference
      at org.wordpress.android.fluxc.store.UploadStore.handleUpdateMedia(UploadStore.java:337)
      at org.wordpress.android.fluxc.store.UploadStore.onMediaAction(UploadStore.java:131)
      at org.wordpress.android.fluxc.store.UploadStore.onAction(UploadStore.java:98)
      at java.lang.reflect.Method.invoke(Native Method)
      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:485)
      at org.greenrobot.eventbus.EventBus.invokeSubscriber(EventBus.java:479) 
      at org.greenrobot.eventbus.AsyncPoster.run(AsyncPoster.java:46) 
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
      at java.lang.Thread.run(Thread.java:761) 
```

--- a/WordPress/src/main/java/org/wordpress/android/ui/main/MeFragment.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/main/MeFragment.java
@@ -55,6 +55,7 @@ import org.wordpress.android.util.HelpshiftHelper.Tag;
 import org.wordpress.android.util.MediaUtils;
 import org.wordpress.android.util.ToastUtils;
 import org.wordpress.android.util.ToastUtils.Duration;
+import org.wordpress.android.util.WPMediaUtils;
 import org.wordpress.android.widgets.WPNetworkImageView;
 
 import java.io.DataInputStream;
@@ -449,17 +450,18 @@ public class MeFragment extends Fragment {
                                     : AnalyticsTracker.Stat.ME_GRAVATAR_GALLERY_PICKED;
                     AnalyticsTracker.track(stat);
                     Uri imageUri = Uri.parse(strMediaUri);
-                    if (!MediaUtils.isInMediaStore(imageUri)) {
-                        // Download the picture. See https://github.com/wordpress-mobile/WordPress-Android/issues/5818
-                        Uri downloadedUri = MediaUtils.downloadExternalMedia(getActivity(), imageUri);
-                        if (downloadedUri != null) {
-                            startCropActivity(downloadedUri);
-                        } else {
-                            ToastUtils.showToast(getActivity(), R.string.error_downloading_image, Duration.SHORT);
+                    if (imageUri != null) {
+                        boolean didGoWell = WPMediaUtils.fetchMediaAndDoNext(getActivity(), imageUri,
+                                new WPMediaUtils.MediaFetchDoNext() {
+                                    @Override
+                                    public void doNext(Uri uri) {
+                                        startCropActivity(uri);
+                                    }
+                                });
+
+                        if (!didGoWell) {
                             AppLog.e(AppLog.T.UTILS, "Can't download picked or captured image");
                         }
-                    } else {
-                        startCropActivity(imageUri);
                     }
                 }
                 break;
@@ -467,7 +469,14 @@ public class MeFragment extends Fragment {
                 AnalyticsTracker.track(AnalyticsTracker.Stat.ME_GRAVATAR_CROPPED);
 
                 if (resultCode == Activity.RESULT_OK) {
-                    fetchMedia(UCrop.getOutput(data));
+                    WPMediaUtils.fetchMediaAndDoNext(getActivity(), UCrop.getOutput(data),
+                            new WPMediaUtils.MediaFetchDoNext() {
+                                @Override
+                                public void doNext(Uri uri) {
+                                    startGravatarUpload(MediaUtils.getRealPathFromURI(getActivity(), uri));
+                                }
+                            });
+
                 } else if (resultCode == UCrop.RESULT_ERROR) {
                     AppLog.e(AppLog.T.MAIN, "Image cropping failed!", UCrop.getError(data));
                     ToastUtils.showToast(getActivity(), R.string.error_cropping_image, Duration.SHORT);
@@ -498,21 +507,6 @@ public class MeFragment extends Fragment {
                 .withAspectRatio(1, 1)
                 .withOptions(options)
                 .start(getActivity(), this);
-    }
-
-    private void fetchMedia(Uri mediaUri) {
-        if (!MediaUtils.isInMediaStore(mediaUri)) {
-            // Do not download the file in async task. See https://github.com/wordpress-mobile/WordPress-Android/issues/5818
-            Uri downloadedUri = MediaUtils.downloadExternalMedia(getActivity(), mediaUri);
-            if (downloadedUri != null) {
-                startGravatarUpload(MediaUtils.getRealPathFromURI(getActivity(), downloadedUri));
-            } else {
-                ToastUtils.showToast(getActivity(), R.string.error_downloading_image, ToastUtils.Duration.SHORT);
-            }
-        } else {
-            // It is a regular local media file
-            startGravatarUpload(MediaUtils.getRealPathFromURI(getActivity(), mediaUri));
-        }
     }
 
     private void startGravatarUpload(final String filePath) {

--- a/WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java
@@ -60,7 +60,6 @@ import org.wordpress.android.util.ActivityUtils;
 import org.wordpress.android.util.AnalyticsUtils;
 import org.wordpress.android.util.AniUtils;
 import org.wordpress.android.util.AppLog;
-import org.wordpress.android.util.CrashlyticsUtils;
 import org.wordpress.android.util.DisplayUtils;
 import org.wordpress.android.util.FluxCUtils;
 import org.wordpress.android.util.ListUtils;
@@ -351,8 +350,13 @@ public class MediaBrowserActivity extends AppCompatActivity implements MediaGrid
             case RequestCodes.VIDEO_LIBRARY:
                 if (resultCode == Activity.RESULT_OK && data != null) {
                     Uri imageUri = data.getData();
-                    String mimeType = getContentResolver().getType(imageUri);
-                    fetchMedia(imageUri, mimeType);
+                    final String mimeType = getContentResolver().getType(imageUri);
+                    WPMediaUtils.fetchMediaAndDoNext(this, imageUri, new WPMediaUtils.MediaFetchDoNext() {
+                        @Override
+                        public void doNext(Uri uri) {
+                            queueFileForUpload(getOptimizedPictureIfNecessary(uri), mimeType);
+                        }
+                    });
                     trackAddMediaFromDeviceEvents(
                             false,
                             requestCode == RequestCodes.VIDEO_LIBRARY,
@@ -707,7 +711,15 @@ public class MediaBrowserActivity extends AppCompatActivity implements MediaGrid
         }
         for (Uri uri : uriList) {
             if (uri != null) {
-                fetchMedia(uri, getContentResolver().getType(uri));
+                WPMediaUtils.fetchMediaAndDoNext(this, uri,
+                        new WPMediaUtils.MediaFetchDoNext() {
+                            @Override
+                            public void doNext(Uri downloadedUri) {
+                                queueFileForUpload(
+                                        getOptimizedPictureIfNecessary(downloadedUri),
+                                        getContentResolver().getType(downloadedUri));
+                            }
+                        });
             }
         }
     }
@@ -827,29 +839,6 @@ public class MediaBrowserActivity extends AppCompatActivity implements MediaGrid
                     WPMediaUtils.launchVideoLibrary(this);
                     break;
             }
-        }
-    }
-
-    private void fetchMedia(Uri mediaUri, final String mimeType) {
-        if (!MediaUtils.isInMediaStore(mediaUri)) {
-            // Do not download the file in async task. See https://github.com/wordpress-mobile/WordPress-Android/issues/5818
-            Uri downloadedUri = null;
-            try {
-                downloadedUri = MediaUtils.downloadExternalMedia(MediaBrowserActivity.this, mediaUri);
-            } catch (IllegalStateException e) {
-                // Ref: https://github.com/wordpress-mobile/WordPress-Android/issues/5823
-                AppLog.e(AppLog.T.UTILS, "Can't download the image at: " + mediaUri.toString(), e);
-                CrashlyticsUtils.logException(e, AppLog.T.MEDIA, "Can't download the image at: " + mediaUri.toString() +
-                        " See issue #5823");
-            }
-            if (downloadedUri != null) {
-                queueFileForUpload(getOptimizedPictureIfNecessary(downloadedUri), mimeType);
-            } else {
-                ToastUtils.showToast(MediaBrowserActivity.this, R.string.error_downloading_image,
-                        ToastUtils.Duration.SHORT);
-            }
-        } else {
-            queueFileForUpload(getOptimizedPictureIfNecessary(mediaUri), mimeType);
         }
     }
 

--- a/WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerActivity.java
+++ b/WordPress/src/main/java/org/wordpress/android/ui/photopicker/PhotoPickerActivity.java
@@ -240,7 +240,14 @@ public class PhotoPickerActivity extends AppCompatActivity
     private void doMediaUriSelected(@NonNull Uri mediaUri, @NonNull PhotoPickerMediaSource source) {
         // if user chose a featured image, we need to upload it and return the uploaded media object
         if (mBrowserType == MediaBrowserType.FEATURED_IMAGE_PICKER) {
-            uploadMedia(mediaUri);
+            WPMediaUtils.fetchMediaAndDoNext(this, mediaUri,
+                    new WPMediaUtils.MediaFetchDoNext() {
+                        @Override
+                        public void doNext(Uri uri) {
+                            uploadMedia(uri);
+                        }
+                    });
+
         } else {
             Intent intent = new Intent()
                     .putExtra(EXTRA_MEDIA_URI, mediaUri.toString())

--- a/WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java
+++ b/WordPress/src/main/java/org/wordpress/android/util/WPMediaUtils.java
@@ -442,4 +442,34 @@ public class WPMediaUtils {
         return ViewConfiguration.get(context).getScaledMaximumFlingVelocity() / 2;
     }
 
+
+    public interface MediaFetchDoNext {
+        void doNext(Uri uri);
+    }
+
+    public static boolean fetchMediaAndDoNext(Context context, Uri mediaUri, MediaFetchDoNext listener) {
+        if (!MediaUtils.isInMediaStore(mediaUri)) {
+            // Do not download the file in async task. See https://github.com/wordpress-mobile/WordPress-Android/issues/5818
+            Uri downloadedUri = null;
+            try {
+                downloadedUri = MediaUtils.downloadExternalMedia(context, mediaUri);
+            } catch (IllegalStateException e) {
+                // Ref: https://github.com/wordpress-mobile/WordPress-Android/issues/5823
+                AppLog.e(AppLog.T.UTILS, "Can't download the image at: " + mediaUri.toString(), e);
+                CrashlyticsUtils.logException(e, AppLog.T.MEDIA, "Can't download the image at: " + mediaUri.toString() +
+                        " See issue #5823");
+            }
+            if (downloadedUri != null) {
+                listener.doNext(downloadedUri);
+            } else {
+                ToastUtils.showToast(context, R.string.error_downloading_image,
+                        ToastUtils.Duration.SHORT);
+                return false;
+            }
+        } else {
+            listener.doNext(mediaUri);
+        }
+        return true;
+    }
+
 }

